egs/wsj/s5/steps/libs/nnet3/xconfig/self_layer.py

```diff
@@ -0,0 +1,218 @@
+        self._left_context = -int(m.group(2))
+        self._input_period = int(m.group(3))
+        self._stats_period = int(m.group(4))
+        self._right_context = int(m.group(5))
+        self._num_heads = self.config['num-heads']
+        self._affine_dim = self.config['affine-dim']
+
+        if self._output_stddev:
+          output_dim = 2 * self.descriptors['input']['dim']
+        else:
+          output_dim = self.descriptors['input']['dim']
+        if self._output_log_counts:
+          output_dim = output_dim + 1
+
+        if self._num_heads > 0:
+          output_dim = output_dim * self._num_heads
+
+
+        if self.config['dim'] > 0 and self.config['dim'] != output_dim:
+            raise RuntimeError(
+                "Invalid dim supplied {0:d} != "
+                "actual output dim {1:d}".format(
+                    self.config['dim'], output_dim))
+        self.config['dim'] = output_dim
+
+    def check_configs(self):
+        if not (self._left_context >= 0 and self._right_context >= 0
+                and self._input_period > 0 and self._stats_period > 0
+                and self._left_context % self._stats_period == 0
+                and self._right_context % self._stats_period == 0
+                and self._stats_period % self._input_period == 0):
+            raise RuntimeError(
+                "Invalid configuration of statistics-extraction: {0}".format(
+                    self.config['config']), self)
+        super(XconfigSelfLayer, self).check_configs()
+
+    def _generate_config(self):
+        input_desc = self.descriptors['input']['final-string']
+        input_dim = self.descriptors['input']['dim']
+
+        # Here, the whole self layer computes a linear combination of a series of frames.
+        # The linear combination weights are computed by a multi layer perceptron.
+        # The multi layer perceptron consists of the following components in sequence:
+        # an affine component, a ReLU component, a batchnorm component, another affine component
+        # and finally a softmax component.
+        # The weights are then append with the same series of frames and pass to 
+        # the SelfAttentionComponent for the computation of linear combination of frames.
+        # The tricky part here is that the softmax computation are done in the SelfAttentionComponent
+        # as the weights of each head are arranged in columns of the CuMatrixBase
+        # We need to transpose the weights to row vector, then call ApplySoftMaxPerRow(), 
+        # then transpose the weights back to column vector.
+
+        da_dim = self._affine_dim
+        self_repair_scale = 1e-05
+        target_rms = 1.0
+        affine_options = 'max-change=0.75'
+
+        configs = []
+        # The first affine node.
+        configs.append(
+            'component name={0}.first_affine type=NaturalGradientAffineComponent'
+            ' input-dim={1} output-dim={2} {3}'
```

Comment:
I am not sure, why do you need to add separate part for having affine+nonlin+batchnorm within this component, when you can add this part using relu-batchnorm-layer at script level.